File: hello-websocket-php/client/ws_client.php
<?php
/**
 * Hello WebSocket Protocol - PHP Client.
 *
 * Implements the full PROTOCOL.md client lifecycle: connect, HELLO/BONJOUR,
 * ping/pong, time notification, kiss request/response, random/hash, disconnect.
 */

require __DIR__ . '/../vendor/autoload.php';

use Ratchet\Client\Connector;
use Ratchet\RFC6455\Messaging\Frame;
use React\EventLoop\Loop;
use HelloWs\Common\ByteWriter;

function sendBinary($conn, string $data): void {
    $conn->send(new Frame($data, true, Frame::OP_BINARY));
}
use function HelloWs\Common\{now_ms, now_iso, log_msg};
use function HelloWs\Common\{encode_frame, decode_message};
use const HelloWs\Common\{
    PORT, CLIENT_LANG,
    MSG_HELLO, MSG_BONJOUR, MSG_ECHO_RESPONSE,
    MSG_KISS_REQUEST, MSG_KISS_RESPONSE, MSG_PING, MSG_PONG,
    MSG_TIME_NOTIFICATION, MSG_RANDOM_NUMBER, MSG_HASH_RESPONSE,
    MSG_DISCONNECT, MSG_ERROR,
    RANDOM_INTERVAL
};

$host = getenv('WS_SERVER') ?: '127.0.0.1';
$port = (int)(getenv('WS_PORT') ?: PORT);

log_msg('ws-client', "Starting PHP WebSocket client [version: 1.0.0]");
$path = getenv('WS_PATH') ?: '/ws';
$url = "ws://$host:$port$path";
log_msg('ws-client', "Connecting to $url");

$loop = Loop::get();
$connected = false;
$maxAttempts = 3;

function tryConnect(string $host, int $port, int $attempt): void
{
    global $loop, $connected, $maxAttempts;
    $path = getenv('WS_PATH') ?: '/ws';
    $url = "ws://$host:$port$path";
    log_msg('ws-client', "Connection attempt $attempt/$maxAttempts to $url");

    $userId = 'php-client-' . now_ms();
    $active = true;

    $connector = new Connector($loop);

    $connector($url, [], ['userId' => $userId])
        ->then(function ($conn) use (&$active, &$connected, $loop, $host, $port, $attempt, $maxAttempts) {
            $connected = true;
            log_msg('ws-client', "Connected");

            // Send HELLO
            $w = new ByteWriter();
            $w->writeString(\HelloWs\Common\CLIENT_LANG);
            sendBinary($conn, encode_frame(\HelloWs\Common\MSG_HELLO, $w->data()));

            // Background: RANDOM_NUMBER every 5s
            $randomId = 1;
            $randomTimer = $loop->addPeriodicTimer(RANDOM_INTERVAL, function () use ($conn, &$active, &$randomId) {
                if (!$active) return;
                try {
                    $num = random_int(PHP_INT_MIN, PHP_INT_MAX);
                    $w = new ByteWriter();
                    $w->writeI64($randomId);
                    $w->writeI64($num);
                    sendBinary($conn, encode_frame(\HelloWs\Common\MSG_RANDOM_NUMBER, $w->data()));
                    log_msg('ws-client', "RANDOM_NUMBER id=$randomId number=$num");
                    $randomId++;
                } catch (\Exception $e) {
                    $active = false;
                }
            });

            // Receive handler
            $conn->on('message', function ($msg) use ($conn, &$active, $loop, $randomTimer) {
                $data = (string)$msg;
                try {
                    $msgObj = decode_message($data);
                } catch (\Exception $e) {
                    log_msg('ws-client', "Decode error: " . $e->getMessage());
                    return;
                }

                switch ($msgObj->type) {
                    case MSG_BONJOUR:
                        log_msg('ws-client', "BONJOUR server_language={$msgObj->serverLanguage}");
                        break;

                    case MSG_PING:
                        $w = new ByteWriter();
                        $w->writeI64($msgObj->timestampMs);
                        sendBinary($conn, encode_frame(\HelloWs\Common\MSG_PONG, $w->data()));
                        break;

                    case MSG_TIME_NOTIFICATION:
                        log_msg('ws-client', "TIME_NOTIFICATION ts={$msgObj->timestampMs} iso={$msgObj->iso8601}");
                        break;

                    case MSG_KISS_REQUEST:
                        log_msg('ws-client', "KISS_REQUEST os={$msgObj->osName} ver={$msgObj->osVersion} rel={$msgObj->osRelease} arch={$msgObj->osArch}");
                        $w = new ByteWriter();
                        $w->writeString(setlocale(LC_ALL, '0') ?: 'en_US');
                        $w->writeString('UTF-8');
                        $w->writeString(date_default_timezone_get());
                        sendBinary($conn, encode_frame(\HelloWs\Common\MSG_KISS_RESPONSE, $w->data()));
                        break;

                    case MSG_ECHO_RESPONSE:
                        log_msg('ws-client', "ECHO_RESPONSE status={$msgObj->echoStatus} results=" . count($msgObj->echoResults ?? []));
                        break;

                    case MSG_HASH_RESPONSE:
                        log_msg('ws-client', "HASH_RESPONSE id={$msgObj->randomId} hash={$msgObj->hashHex}");
                        break;

                    case MSG_ERROR:
                        log_msg('ws-client', "ERROR code={$msgObj->errorCode} msg={$msgObj->errorMessage}");
                        break;

                    case MSG_DISCONNECT:
                        log_msg('ws-client', "DISCONNECT reason={$msgObj->disconnectReason}");
                        $active = false;
                        $loop->cancelTimer($randomTimer);
                        $conn->close();
                        break;

                    default:
                        $hex = sprintf('0x%02x', $msgObj->type);
                        log_msg('ws-client', "Unknown message type: $hex");
                        break;
                }
            });

            $conn->on('close', function () use (&$active, &$connected, $loop, $randomTimer, $host, $port, $attempt, $maxAttempts) {
                if ($active) {
                    log_msg('ws-client', "Disconnected");
                }
                $active = false;
                $connected = false;
                try { $loop->cancelTimer($randomTimer); } catch (\Exception $e) {}
                // Schedule retry if we haven't exceeded max attempts
                if ($attempt < $maxAttempts) {
                    $loop->addTimer(2.0, function () use ($host, $port, $attempt) {
                        tryConnect($host, $port, $attempt + 1);
                    });
                } else {
                    log_msg('ws-client', "Failed to connect after $maxAttempts attempts");
                    $loop->stop();
                }
            });

            $conn->on('error', function ($e) use (&$active, $loop, $randomTimer) {
                log_msg('ws-client', "Exception: " . $e->getMessage());
                $active = false;
                try { $loop->cancelTimer($randomTimer); } catch (\Exception $ex) {}
            });
        })
        ->otherwise(function ($e) use ($attempt, $maxAttempts, $loop, $host, $port) {
            log_msg('ws-client', "Error: " . $e->getMessage());
            if ($attempt < $maxAttempts) {
                $loop->addTimer(2.0, function () use ($host, $port, $attempt) {
                    tryConnect($host, $port, $attempt + 1);
                });
            } else {
                log_msg('ws-client', "Failed to connect after $maxAttempts attempts");
                $loop->stop();
            }
        });
}

tryConnect($host, $port, 1);

$loop->run();

File: hello-websocket-cpp/client/ws_client.cpp
// Hello WebSocket Protocol - C++ Client
// Implements the full PROTOCOL.md client lifecycle: connect, HELLO/BONJOUR,
// ping/pong, time notification, kiss request/response, random/hash, disconnect.
#include "../common/codec.hpp"
#include "../common/ws.hpp"
#include <thread>
#include <mutex>
#include <atomic>
#include <sstream>
#include <cstring>
#include <cstdlib>

using namespace hws;

bool tryConnect(const std::string& host, uint16_t port);

int main() {
    const char* hostEnv = std::getenv("WS_SERVER");
    std::string host = hostEnv ? hostEnv : "127.0.0.1";

    const char* portEnv = std::getenv("WS_PORT");
    uint16_t p = PORT;
    if (portEnv) {
        int v = atoi(portEnv);
        if (v > 0) p = static_cast<uint16_t>(v);
    }

    log("ws-client", "Starting C++ WebSocket client [version: 1.0.0]");
    std::string url = "ws://" + host + ":" + std::to_string(p);
    log("ws-client", "Connecting to " + url);

    for (int attempt = 1; attempt <= 3; attempt++) {
        log("ws-client", "Connection attempt " + std::to_string(attempt) + "/3 to " + url);
        if (tryConnect(host, p)) return 0;
        if (attempt < 3) std::this_thread::sleep_for(std::chrono::seconds(2));
    }

    log("ws-client", "Failed to connect after 3 attempts");
    return 1;
}

bool tryConnect(const std::string& host, uint16_t port) {
    WSClient ws;
    std::string userId = "cpp-client-" + std::to_string(nowMs());

    if (!ws.connect(host, port, userId)) {
        log("ws-client", "Connection failed");
        return false;
    }

    log("ws-client", "Connected");

    std::mutex sendMutex;
    std::atomic<bool> active(true);

    // Helper: send masked binary frame (client-to-server must be masked)
    auto sendFrame = [&](const std::vector<uint8_t>& data) -> bool {
        std::lock_guard<std::mutex> lock(sendMutex);
        bool result = false;
        if (active) result = wsSendBinaryMasked(ws.sock, data);
        return result;
    };

    // Send HELLO
    {
        Message m{};
        m.type = MSG_HELLO;
        m.clientLanguage = CLIENT_LANG;
        sendFrame(m.encode());
    }

    // Background: RANDOM_NUMBER every 5s
    std::thread randomThread([&]() {
        std::mt19937_64 rng(std::random_device{}());
        int64_t randomId = 1;
        while (active) {
            std::this_thread::sleep_for(std::chrono::milliseconds(RANDOM_INTERVAL_MS));
            if (!active) break;
            int64_t num = static_cast<int64_t>(rng());
            Message m{};
            m.type = MSG_RANDOM_NUMBER;
            m.randomId = randomId;
            m.randomNumber = num;
            if (!sendFrame(m.encode())) { active = false; break; }
            log("ws-client", "RANDOM_NUMBER id=" + std::to_string(randomId) + " number=" + std::to_string(num));
            randomId++;
        }
    });

    // Receive loop
    try {
        while (active) {
            WSFrame frame = wsRecvFrame(ws.sock, false);
            if (!frame.isValid || frame.isClose) break;

            Message msg;
            try {
                msg = decodeMessage(frame.payload.data(), frame.payload.size());
            } catch (const std::exception& e) {
                log("ws-client", std::string("Decode error: ") + e.what());
                continue;
            }

            switch (msg.type) {
                case MSG_BONJOUR:
                    log("ws-client", "BONJOUR server_language=" + msg.serverLanguage);
                    break;

                case MSG_PING:
                    {
                        Message pong{};
                        pong.type = MSG_PONG;
                        pong.timestampMs = msg.timestampMs;
                        sendFrame(pong.encode());
                    }
                    break;

                case MSG_TIME_NOTIFICATION:
                    log("ws-client", "TIME_NOTIFICATION ts=" + std::to_string(msg.timestampMs) + " iso=" + msg.iso8601);
                    break;

                case MSG_KISS_REQUEST:
                    log("ws-client", "KISS_REQUEST os=" + msg.osName + " arch=" + msg.osArch);
                    {
                        Message resp{};
                        resp.type = MSG_KISS_RESPONSE;
                        resp.kissLanguage = "en_US";
                        resp.kissEncoding = "UTF-8";
                        resp.kissTimeZone = "UTC";
                        sendFrame(resp.encode());
                    }
                    break;

                case MSG_ECHO_RESPONSE:
                    log("ws-client", "ECHO_RESPONSE status=" + std::to_string(msg.echoStatus) + " results=" + std::to_string(msg.echoResults.size()));
                    for (size_t i = 0; i < msg.echoResults.size(); i++) {
                        const auto& r = msg.echoResults[i];
                        std::string kvStr;
                        bool first = true;
                        for (const auto& [k, v] : r.kv) {
                            if (!first) kvStr += ", ";
                            kvStr += k + "=" + v;
                            first = false;
                        }
                        log("ws-client", "  Result #" + std::to_string(i + 1) + ": idx=" + std::to_string(r.idx) + " type=" + std::to_string(r.type) + " kv={" + kvStr + "}");
                    }
                    break;

                case MSG_HASH_RESPONSE:
                    log("ws-client", "HASH_RESPONSE id=" + std::to_string(msg.randomId) + " hash=" + msg.hashHex);
                    break;

                case MSG_ERROR:
                    log("ws-client", "ERROR code=" + std::to_string(msg.errorCode) + " msg=" + msg.errorMessage);
                    break;

                case MSG_DISCONNECT:
                    log("ws-client", "DISCONNECT reason=" + msg.disconnectReason);
                    active = false;
                    break;

                default: {
                    std::ostringstream ss;
                    ss << std::hex << static_cast<int>(msg.type);
                    log("ws-client", "Unknown message type: 0x" + ss.str());
                }
            }
        }
    } catch (const std::exception& e) {
        log("ws-client", std::string("Exception: ") + e.what());
    } catch (...) {
        log("ws-client", "Unknown exception");
    }

    // Cleanup: always join thread before returning
    active = false;
    if (randomThread.joinable()) randomThread.join();

    // Send DISCONNECT
    {
        Message m{};
        m.type = MSG_DISCONNECT;
        m.disconnectReason = "client shutdown";
        sendFrame(m.encode());
    }

    wsSendControl(ws.sock, 0x08, {}, true);
    ws.close();
    log("ws-client", "Disconnected");
    return true;
}
